### PR TITLE
ST2 Pack for GreyNoise API

### PR DESCRIPTION
**Pull request by bradchiappetta**
This is a pack with actions to interact with the GreyNoise API.  More information at greynoise.io/tech

Would love some feedback on this as this is my first pack submission, and I'm sure there are plenty of places for improvement.

### CODE REVIEW

File: greynoise/actions/workflows/ip_lookup.yaml
Comment (blag) [inline greynoise/actions/workflows/ip_lookup.yaml:30]: These are dictionary keys with no corresponding values, which isn't valid YAML.

I believe the correct YAML and Orquesta syntax is to make them elements of a list.
```suggestion
          - result=<% result().result %>
          - is_riot=true
```

File: greynoise/actions/workflows/ip_lookup.yaml
Comment (blag) [inline greynoise/actions/workflows/ip_lookup.yaml:35]:
```suggestion
          - result=<% result().stderr %>
```

File: greynoise/actions/workflows/ip_lookup.yaml
Comment (blag) [inline greynoise/actions/workflows/ip_lookup.yaml:38]:
```suggestion
          - result=<% result().result %>
```

File: greynoise/actions/workflows/ip_lookup.yaml
Comment (blag) [inline greynoise/actions/workflows/ip_lookup.yaml:48]:
```suggestion
          - result=<% result().result[0] %>
```

File: greynoise/actions/workflows/ip_lookup.yaml
Comment (blag) [inline greynoise/actions/workflows/ip_lookup.yaml:44]: If the action successfully returns but `result().result` is empty, this will cause a YAQL error, which is more difficult to debug. I would add a length check in this.

File: greynoise/actions/workflows/ip_lookup.yaml
Comment (blag) [inline greynoise/actions/workflows/ip_lookup.yaml:58]:
```suggestion
          - is_riot=false
          - is_seen=true
          - result=<% result().result %>
```

File: greynoise/config.schema.yaml
Comment (blag) [inline greynoise/config.schema.yaml:4]: I would include some text about the default value will be pulled from the `gn_api_key` key in key-value store.

File: greynoise/config.schema.yaml
Comment (blag) [inline greynoise/config.schema.yaml:8]: I would rename the key to something more explicit. If you're an ST2 admin unfamiliar with GreyNoise, and you're listing st2kv keys when you run across one named `gn_api_key`, are you going to know what it's for?

A better name is probably `greynoise_api_key`.

```suggestion
    default: "{{ st2kv.system.greynoise_api_key }}"
```

File: greynoise/docs/.gitignore
Comment (blag) [inline greynoise/docs/.gitignore:1]: You may want to rename this file to `.gitkeep` instead.

File: greynoise/greynoise.yaml.example
Comment (blag) [inline greynoise/greynoise.yaml.example:3]: This is an example config value, and it will only be explicitly set if it is overriding the `gn_api_key` key from the key-value store, so setting it explicitly to pull from that in the example gives your users the wrong impression.

I would make this something that is obviously fake, like:

```suggestion
gn_api_key: GREYNOISE_SECRET_API_KEY
```

File: greynoise/pack.yaml
Comment (blag) [inline greynoise/pack.yaml:7]: I wish all pack descriptions were this clear about the services they integrate with - good job! 😄

File: greynoise/LICENSE
Comment (blag) [inline greynoise/LICENSE:1]: NAK - I appreciate the liberal license, but all new packs in Exchange MUST be licensed under the Apache 2 License. That way all users who can install StackStorm itself know that they can also install any pack on the StackStorm Exchange (although I realize that the MIT license is somewhat less restrictive than the Apache license, consistency make everybody's lives easier in this regard).

There are two packs on Exchange that have been grandfathered into this policy, which I think is probably where you got the impression that MIT was okay. Both of those packs have PRs changing their licensing, but both are held up by contributors who haven't accepted the PR changes (and unfortunately may never do so).

**Review by blag — CHANGES_REQUESTED** [review]
This is good initial work, it's looking very good for a first pass!

There are some small changes to make, and one potential bug to fix, and the license needs to be changed, but otherwise this is looking good. Thank you for the contribution!

File: greynoise/actions/lib/base.py
Comment (bradchiappetta) [reply greynoise/actions/lib/base.py:13]: @blag : I agree, I'm not thrilled with hardcoding this value, but we use this for tracking metrics on our side, so desire to keep a version here.  Is there a way that I can dynamically pull this is from the client when it is initialized?  I'll at least move it out to a variable for now, so it is more obvious, but if there is a way to pull it from the pack.yaml, that would be perfect.

Comment (blag) [reply greynoise/actions/lib/base.py:13]: I'm not aware of a way to pull it from the `pack.yaml`. 😕

I understand the "need" to track for metrics, but I'm not so sure that it's going to have as much value as you'd like.

Consider the following: we release this pack as v1.0.0, and then at some point the pack gets revved to v1.0.1, and then v1.1.0, and then 1.2.0. But since this pack is a bit of a unicorn in this regard, the version here is not updated, so all of those pack revisions will still use the `greynoise-stackstorm-v1.0.0` integration name.

In that case, you will have _worse_ data, because it will be lying to you. People who are using v1.0.1, v1.1.0, and v1.2.0 will all show up in your metrics as using the v1.0.0 of this pack.

Instead, I would just use the integration name `greynoise-stackstorm`. It doesn't include a version name, sure, but at least your metrics are no longer lying to you.

Alternatively, you could add an immutable key to the pack config schema <sup>1</sup>, with the default value set to the (current) pack value. Pull the version from the pack config in your code, and drop a YAML comment next to or on the same line as the version in `pack.yaml` to please also update the version in the pack config schema:

`pack.yaml`:
```yaml
...
version: 1.0.0  # Please also update the version in the config.schema.yaml!
...
```

That is by no means bulletproof, but if you're lucky your metrics will be slightly more accurate.

I did just have one idea, let me bounce this off of you: GreyNoise can donate money, and we would provide you with the raw metrics from GitHub. We would have to make sure that we can get the metrics from GitHub with the detail that would be valuable to you, but that might be a win-win. StackStorm has been searching for ways to get more community contributions, and that might be a good way. Now, of course, I/we would have to discuss this with the TSC and get approval for this, so I can't promise anything here and now.

As of now, to my knowledge, you are the only person to have requested this feature of StackStorm. If this is a common use-case for integration packs, then I think we would be amenable to a PR to ST2 that made it easier to access values from `pack.yaml` files.

<sup>1</sup> However, I would _not_ include the key in the config example, because you don't want to allow users to override it anyway. And you don't want to hint at it, because if the value is marked as immutable in the config schema, then they won't be able to override it anyway.

File: greynoise/actions/workflows/ip_lookup.yaml
Comment (bradchiappetta) [reply greynoise/actions/workflows/ip_lookup.yaml:30]: @blag : So, this is working before this suggested change (the workflow outputs what I expect) but when I tried making this change,  the workflow fails with:

```
{
  "errors": [
    {
      "expression": "<% succeeded() and results()|length >=1 and result().result[0].noise = true %>",
      "message": "Lexical error: illegal character '|' at position 25",
      "spec_path": "tasks.check_quick.next[0].when",
      "schema_path": "properties.tasks.patternProperties.^\\w+$.properties.next.items.properties.when",
      "language": "yaql",
      "type": "expression"
    },
    {
      "expression": "<% succeeded() and results()|length >=1 and result().result[0].noise = false %>",
      "message": "Lexical error: illegal character '|' at position 25",
      "spec_path": "tasks.check_quick.next[1].when",
      "schema_path": "properties.tasks.patternProperties.^\\w+$.properties.next.items.properties.when",
      "language": "yaql",
      "type": "expression"
    },
    {
      "message": "['result=<% result().result[0] %>'] is not valid under any of the given schemas",
      "spec_path": "tasks.check_quick.next[1].publish",
      "schema_path": "properties.tasks.patternProperties.^\\w+$.properties.next.items.properties.publish.oneOf",
      "type": "syntax"
    },
    {
      "message": "['result=<% result().result %>', 'is_riot=true'] is not valid under any of the given schemas",
      "spec_path": "tasks.check_riot.next[0].publish",
      "schema_path": "properties.tasks.patternProperties.^\\w+$.properties.next.items.properties.publish.oneOf",
      "type": "syntax"
    },
    {
      "message": "['result=<% result().stderr %>'] is not valid under any of the given schemas",
      "spec_path": "tasks.check_riot.next[2].publish",
      "schema_path": "properties.tasks.patternProperties.^\\w+$.properties.next.items.properties.publish.oneOf",
      "type": "syntax"
    },
    {
      "message": "['result=<% result().result %>'] is not valid under any of the given schemas",
      "spec_path": "tasks.check_riot.next[3].publish",
      "schema_path": "properties.tasks.patternProperties.^\\w+$.properties.next.items.properties.publish.oneOf",
      "type": "syntax"
    },
    {
      "message": "['is_riot=false', 'is_seen=true', 'result=<% result().result %>'] is not valid under any of the given schemas",
      "spec_path": "tasks.get_context.next[0].publish",
      "schema_path": "properties.tasks.patternProperties.^\\w+$.properties.next.items.properties.publish.oneOf",
      "type": "syntax"
    }
  ],
  "output": null
}
```

File: greynoise/actions/context_ip.yaml
Comment (bradchiappetta) [reply greynoise/actions/context_ip.yaml]: Removed for clarity.

File: greynoise/actions/lib/base.py
Comment (bradchiappetta) [reply greynoise/actions/lib/base.py]: updated to gn_client

File: greynoise/actions/context_ip.py
Comment (bradchiappetta) [reply greynoise/actions/context_ip.py]: Reworked to not use a variable here.

File: greynoise/actions/lib/base.py
Comment (bradchiappetta) [reply greynoise/actions/lib/base.py]: Per the suggestion from @nmaludy : I was able to get this by doing a read/regex from the pack.yaml file, so this is now dynamically pulled whenever it is updated.  Thanks for the suggestions here.

File: greynoise/actions/ip_lookup.yaml
Comment (bradchiappetta) [reply greynoise/actions/ip_lookup.yaml]: Renamed to ip_lookup just in case.

File: greynoise/actions/workflows/ip_lookup.yaml
Comment (bradchiappetta) [reply greynoise/actions/workflows/ip_lookup.yaml]: added length check.

File: greynoise/config.schema.yaml
Comment (bradchiappetta) [reply greynoise/config.schema.yaml]: Updated.

File: greynoise/config.schema.yaml
Comment (bradchiappetta) [reply greynoise/config.schema.yaml]: updated.

File: greynoise/LICENSE
Comment (bradchiappetta) [reply greynoise/LICENSE]: Updated.  Please validate I used the correct one now.

**Review by blag — APPROVED** [review]
:+1: I didn't even think about doing that.

**Review by blag — CHANGES_REQUESTED** [review]
Still need to fix the publishing syntax, but otherwise looks good.

File: greynoise/actions/workflows/ip_lookup.yaml
Comment (bradchiappetta) [reply greynoise/actions/workflows/ip_lookup.yaml:30]: @blag : I'm confused, this exact SYNTAX is what is in the current version?  And during my testing, it works as expected?

**Comment by bradchiappetta** [comment]
> Still need to fix the publishing syntax, but otherwise looks good.

@blag : so, when I tried your suggested modification, that is when I got the stacktrace error (with the changes to publishing).  If I run this as it is currently in the PR, it all works as expected with no errors.  Can you provide some additional guidance?

File: greynoise/actions/workflows/ip_lookup.yaml
Comment (blag) [inline greynoise/actions/workflows/ip_lookup.yaml:52]: I'm still not sure this is what you want - need to check `result().result.len()`

```suggestion
      - when: >-
          <% succeeded() and
          result().result.len() >=1 and
          result().result[0].noise = true %>
        do: get_context
      - when: >-
          <% succeeded() and
          result().result.len() >=1 and
          result().result[0].noise = false %>
```

Comment (bradchiappetta) [reply greynoise/actions/workflows/ip_lookup.yaml]: Ah, yes, I see the difference, you are absolutely right.

**Review by blag — APPROVED** [review]

**Comment by blag** [comment]
I think this looks good. The Orquesta context variable publishing still looks a little weird to me, but if it's working for you then that must be a supported syntax.

Thank you for the contribution!

I'll be releasing v3.4 tomorrow, so I won't have time to put this on Exchange until Wednesday at the earliest. Feel free to ping me if I forget.

**Comment by bradchiappetta** [comment]
@blag : Congrats on getting ST2 v3.4 shipped! Just wanted to remind you about this, as you suggested.  Thanks!

**Comment by blag** [comment]
I have a few 3.4.1 post-release items to take care of, but I should get to this later today/tonight, or tomorrow. :+1:

**Comment by blag** [comment]
Sorry it took me so long. I have created the GreyNoise pack here:
https://github.com/StackStorm-Exchange/stackstorm-greynoise

I still need to setup the Slack integration, but I'm going to need some help with that and I don't consider it a blocker.

I have invited you as a collaborator with write access, which should help you maintain the code. We try to create pull requests for every change we make to packs, and we follow SemVer to let downstream users (which may be other pack authors) know when certain versions have breaking changes.

While you are still starting out, please create pull requests and request a review from @nmaludy, me, or @StackStorm-Exchange/stackstorm-exchange, and wait to merge until somebody has signed off on it. But as you get more comfortable with maintaining the pack, you don't need to wait for reviews all of the time (although please do still put your changes into a PR). I'll promote you to have Maintainer access to this repository once you've got a few pack versions published.

If you need anything else, there are some great people in the `#exchange` channel in our [Slack community](https://stackstorm-community.slack.com), feel free to stop by to ask questions, say hi, or just hang out.

Closing.

**Pull request closed without merging**
